python/cudf/cudf/core/buffer/buffer.py

```diff
@@ -0,0 +1,298 @@
+
+    Parameters
+    ----------
+    ptr : int
+        An integer representing a pointer to device memory.
+    size : int, optional
+        Size of device memory in bytes.
+    owner : object, optional
+        Python object to which the lifetime of the memory allocation is tied.
+        A reference to this object is kept in the returned wrapper object.
+
+    Return
+    ------
+    SimpleNamespace
+        An object that exposes `__cuda_array_interface__` and keeps a reference
+        to `owner`.
+    """
+
+    if size < 0:
+        raise ValueError("size cannot be negative")
+
+    return SimpleNamespace(
+        __cuda_array_interface__={
+            "data": (ptr, False),
+            "shape": (size,),
+            "strides": None,
+            "typestr": "|u1",
+            "version": 0,
+        },
+        owner=owner,
+    )
+
+
+class Buffer(Serializable):
+    """A Buffer represents device memory.
+
+    Use the factory function `as_buffer` to create a Buffer instance.
+    """
+
+    _ptr: int
+    _size: int
+    _owner: object
+
+    def __init__(self):
+        raise ValueError(
+            f"do not create a {self.__class__} directly, please "
+            "use the factory function `cudf.core.buffer.as_buffer`"
+        )
+
+    @classmethod
+    def _from_device_memory(cls: Type[T], data: Any) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = cls.__new__(cls)
+        ret._owner = data
+        if isinstance(data, rmm.DeviceBuffer):  # Common case shortcut
+            ret._ptr = data.ptr
+            ret._size = data.size
+        else:
+            ret._ptr, ret._size = get_ptr_and_size(
+                data.__cuda_array_interface__
+            )
+        if ret.size < 0:
+            raise ValueError("size cannot be negative")
+        return ret
+
+    @classmethod
+    def _from_host_memory(cls: Type[T], data: Any) -> T:
+        """Create a Buffer from a buffer or array like object
+
+        Data must implement `__array_interface__`, the buffer protocol, and/or
+        be convertible to a buffer object using `numpy.asarray()`
+
+        The host memory is copied to a new device allocation.
+
+        Parameters
+        ----------
+        data : array-like or buffer-like
+            An object that represens host memory.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing a copy of `data`.
+        """
+
+        # Extract pointer and size
+        ptr, size = get_ptr_and_size(numpy.asarray(data).__array_interface__)
```

Comment:
Just to be sure: are there any situations in which `numpy.asarray()` creates an unnecessary copy of `data`? Also, is it useful to use `ascontiguousarray()` here and not raise in `get_ptr_and_size`?

Reply:
> Just to be sure: are there any situations in which numpy.asarray() creates an unnecessary copy of data?

No unnecessary, but an object itself might make a copy when exposing its array or buffer interface. Also, objects like `list` will be copied. Anyway, this is a good point since we need to keep the converted numpy array alive until _after_ the copy to device memory. Fixed.

> Also, is it useful to use `ascontiguousarray()` here and not raise in `get_ptr_and_size`?

I am not sure, if the buffer isn't contiguous, it is probably a bug? Might be best to force the user to convert the input manually before calling `as_buffer()`?